### ISSUE TITLE
Competition between thingShadows generated with same client id but different certificates

### ISSUE THREAD

**Issue by wjy617** [aws/aws-iot-device-sdk-js#48]
I have created two different thingShadows with two different groups of keyPath, certPath, caPath and region (both generated from same AWS account), but with same clientId. Each one works well individually, but when they start at the same time, they will compete with each other and both show "connection lost - will attempt reconnection". Is this normal? If not, how can I fix this problem? 

Thanks

**Comment by bhadrip** [comment]
Hello @wjy617 ,

Yes it is expected behavior. Please refer to [this](http://docs.aws.amazon.com/iot/latest/developerguide/protocols.html) documentation. 

``` 
The message broker uses the client ID to identify each client. The client ID is passed in from the client to the message broker as part of the MQTT payload. Two clients with the same client ID are not allowed to be connected concurrently to the message broker. When a client connects to the message broker using a client ID that another client is using, a CONNACK message will be sent to both clients and the currently connected client will be disconnected.
```

Hope this helps.

Thanks
Bhadri

**Comment by wjy617** [comment]
@bhadrip 

Thanks Bhadir, but is there a security issue here, if two different certificate don't have separate mqtt broker? Take this example:

User 1 with certificate 1 that needs to download thing shadow 1. User 2 with certificate 2 that needs to download thing shadow 2. Now, if user 2 can guess the connection ID of user 1, they can interrupt the mqtt connection of user 1. The certificate alone should secure the communication channel right? In this case, the communication channel has client ID as an additional security weakness.

**Comment by bhadrip** [comment]
Hello @wjy617 ,

Policies takes care of the security issue you illustrated. 

Even if user 2 can guess the connection ID of user 1, the policy attached to user 2's certificate will not have permissions to connect to AWS IoT using connection ID 1. The policy will look like this,

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iot:Connect"
            ],
            "Resource": [
                "arn:aws:iot:us-east-1:123456789012:client/ConnectionID1"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "iot:Publish",
                "iot:Subscribe",
                "iot:Receive"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

More about policies [here](http://docs.aws.amazon.com/iot/latest/developerguide/authorization.html).

Does this answer your concern ?

Thanks
Bhadri

**Comment by wjy617** [comment]
Hi @bhadrip,

Thanks for your reply and it answers my question perfectly. Here I have one more question related to clientId, the [document](http://docs.aws.amazon.com/iot/latest/developerguide/iot-limits.html) indicates the client id size limit is 128bytes in UTF-8, but I found the limit is only 62bytes in UTF-8 on my end, id longer than this will cause connection timeout. Could you help me with this ?

**Comment by bhadrip** [comment]
Hi @wjy617 ,

Just wanted to emphasize that it is 128 bytes and not 128 UTF-8 characters(which could be all the way up to 128*4 bytes). can you share the Client Id ?

Thanks
Bhadri

**Comment by wjy617** [comment]
Hi @bhadrip ,

The client id is like "4ce4a6580223f7f24ea372595e46352de01148b5641271e900298b26f141", which has 64 characters and encoded in UTF-8

Thanks

**Comment by bhadrip** [comment]
Hi @wjy617 ,

I am able to connect the example program(`device-example.js`) using the provided client id. The command is 
```node device-example.js -g us-east-1 -H <your unique id>.iot.us-east-1.amazonaws.com -c ./cetificate.crt -a ./root-CA.crt -k ./809467efa6-private.pem.key -i 4ce4a6580223f7f24ea372595e46352de01148b5641271e900298b26f141```

How did you encode the client-id into UTF-8 ? Can you share the code ?

Thanks
Bhadri

**Comment by wjy617** [comment]
Hi @bhadrip, 

I am using webstorm to code, and it is encoded in UTF-8. I also tried to use[ utf8 package](https://www.npmjs.com/package/utf8) to force encoding the clientId string with UTF-8, but it doesn't work neither. I found when I reduce the client id no more than 62 characters, it will work.  The clientId I choose is  the "certificate_id" returned from AWS-iot, does this affect the result?

Thanks

**Comment by bhadrip** [comment]
Hi @wjy617 ,

Can you please share the code snippet that sets the Client ID ? I was able to connect with the client id that you provided. It could also be helpful if you can take a look at the [example ](https://github.com/aws/aws-iot-device-sdk-js/blob/master/examples/device-example.js) if not checked already. It's a good place to start.

Thanks
Bhadri

**Comment by wjy617** [comment]
Hi @bhadrip , 

I have tried device-example.js using commands
`node device-example.js -g us-east-1 -H <my client id>.iot.us-east-1.amazonaws.com -c ./certificate.pem.crt -a ./root-CA.crt -k ./private.pem.key -i 4ce4a6580223f7f24ea372595e46352de01148b5641271e900298b26f141`, 

and got error 
```
error { [Error: getaddrinfo ENOTFOUND <my client id>.iot.us-east-1.amazonaws.com <my client id>.iot.us-east-1.amazonaws.com:8883]
  code: 'ENOTFOUND',
  errno: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: '<my client id>.iot.us-east-1.amazonaws.com',
  host: '<my client id>.iot.us-east-1.amazonaws.com',
  port: 8883 }
offline
connection lost - will attempt reconnection in 4 seconds...
close
```

And I have tried another iotTest.js,
```
var awsIot = require('aws-iot-device-sdk');
var keyPathPrivate = '../awsCerts/private.pem.key';
var certPath = '../awsCerts/certificate.pem.crt';
var caPath = '../awsCerts/root-CA.crt';
var region = 'us-east-1';

function test() {
    var clientId = makeid();
    var iotconfigParam = {
        'keyPath'  : keyPathPrivate,
        'certPath' : certPath,
        'caPath'   : caPath,
        'region'   : region,
        'clientId' : clientId
    };
    var connection = awsIot.thingShadow(iotconfigParam);
    connection.register('003');

    setInterval(function(){
        connection.update('003', {"state": {"reported": {"timestamp": new Date()}}});
    }, 10);

    connection.on('connect', function(){
        console.log('connnected!');
    });

    connection.on('timeout', function(){
        console.log('timeout');
    });

    connection.on('status', function(thingName, stat, clientToken, stateObject){
        console.log("thingName: " + thingName + "\n stat: " + stat + "\n clientToken: " + clientToken + "\n stateObject: " + JSON.stringify(stateObject));
    });

    connection.on('reconnect', function(){
        console.log('reconnect');
    });

    connection.on('disconnect', function(){
        console.log('disconnect');
    });
}

function makeid()
{
    var text = "";
    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

    for( var i=0; i < 62; i++ )
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    console.log(text);
    return text;
}

test();
```

Interestingly, when I generate a clientId longer than  62, it will time out immediately, but when clientId length = 62, connection will timeout after 10 time connection.update() operation;  when clientId length = 61, connection will timeout after 100 time connection.update() operation;  when clientId length = 60, connection will timeout after 1000 time connection.update() operation;  when clientId length = 59, connection will timeout after 10000 time connection.update() operation.

Could you give a hint on this?

Thanks

**Comment by wjy617** [comment]
Hi @bhadrip,

An update: I have tried  device-example.js again with another id, which turns to be the right one, and it  
                  works with the long client id now.
                  But for iotTest.js test, it still fails with client Id longer than 62 on my end, do you have any  
                  idea?

**Comment by wjy617** [comment]
Hi @bhadrip,

An update: I have tried device-example.js again with another id, which turns to be the right one, and it
works with the long client id now.
But for iotTest.js test, it still fails with client Id longer than 62 on my end, do you have any idea?

**Comment by gkwicker** [comment]
Hi @wjy617, we are able to duplicate the issue you observed here.  The root cause is that shadow operations have a client token associated with them which is generated from the client ID and a running integer count.  The maximum length for this client token is 64 bytes.  So, to use the shadow APIs, you'll want to make sure that your client ID is considerably less than 64 bytes to allow for a large number of client token values.  This restriction on client ID length doesn't exist when using the MQTT publish/subscribe APIs.  We will look into workarounds for the shadow API client ID length limitation in a future release.  Thanks for letting us know about this, and thanks for using AWS IoT.

**Comment by wjy617** [comment]
Hi @gkwicker , thanks for your response

**Comment by liuszeng** [comment]
Addressed in release [v1.0.13](https://github.com/aws/aws-iot-device-sdk-js/releases/tag/v1.0.13).

**Comment by greenbreakfast** [comment]
Hi @bhadrip @gkwicker 
Regarding the original question: @bhadrip mentioned:

> Even if user 2 can guess the connection ID of user 1, the policy attached to user 2's certificate will not have permissions to connect to AWS IoT using connection ID 1.

I understand this, but won't User 1's (valid) connection be broken by User 2's connection? In this case, User 2 can disrupt all communication to User 1's device.
I found a similar discussion (https://forums.aws.amazon.com/thread.jspa?threadID=227279) and it suggests creating two policies, one generic for shadow operations and one specific to each certificate. 
Doesn't that defeat the purpose of the generic policy? It would be really nice to have many devices, each with its own certificate that is attached to just a generic policy (that takes care of this connection loop hole). 

Any insights would be appreciated.
